unicore/hub/service/validators.py

```diff
@@ -1,25 +1,87 @@
+import re
+
 import colander
 
+from unicore.hub.service.models import App as AppModel, User as UserModel
+from unicore.hub.service.utils import translation_string_factory as _
+
+
+""" User schema validators
+"""
+
 
-def username_validator(node, value):
+# exclude control characters
+USERNAME_DISALLOWED_CHARS = u'\u0000-\u001f\u007f-\u009f'
+# have a look at http://unicode.org/reports/tr20/#Suitable
+# only excluding bidi controls
+USERNAME_DISALLOWED_CHARS += u'\u202a-\u202e'
+USERNAME_DISALLOWED_CHARS = re.compile(u'[%s]' % USERNAME_DISALLOWED_CHARS)
+
+
+def username_char_validator(node, value):
     # http://stackoverflow.com/questions/1597743/what-restrictions-should-i-impose-on-usernames
-    pass  # TODO - must raise colander.Invalid
+    mapping = {'username': value}
+
+    # disallow leading and trailing space
+    if value[0] == ' ':
+        raise colander.Invalid(
+            node, _('${username} has leading space', mapping=mapping))
+    if value[-1] == ' ':
+        raise colander.Invalid(
+            node, _('${username} has trailing space', mapping=mapping))
+
+    # disallow more than one space in a row
+    if '  ' in value:
+        raise colander.Invalid(
+            node,
+            _('${username} has more than 1 space in a row', mapping=mapping))
+
+    if USERNAME_DISALLOWED_CHARS.search(value):
+        raise colander.Invalid(
+            node,
+            _('${username} contains control characters', mapping=mapping))
+
+
+username_length_validator = colander.Length(
+    min=1, max=UserModel.username_length)
+
+
+@colander.deferred
+def username_validator(node, kw):
+    request = kw.get('request')
+
+    def username_unique_validator(node, value):
+        user = request.db.query(UserModel) \
+            .filter(UserModel.username == value) \
+            .first()
```

Comment:
I actually picked this up from another SQLAlchemy project. I know you and @miltontony aren't fond of using `\` for line continuation. I like writing one query mutation per line for clarity but I'm happy to change it.